Add devicetree support to Rust on Zephyr

d3zd3z opened the pull request:
This is the beginnings of adding support for the devicetree to Rust on Zephyr. This currently:

- Decodes the generated `zephyr.dts` and the `devicetree_generated.h` header, and builds a representation of that DT in memory at build time.
- Outputs the DT as a series of Rust modules. Currently support is fairly minimal, but a few things work:
  - The entire node tree is represented as a tree of modules.
  - Direct phandle references result in new modules that "pub use" the contents of their target. This effectively means, for example, Rust code can reference `zephyr::devicetree::aliases::led0` and get the same node as the actual led node this alias points to. Deeper nested phandles are not yet resolved.
  - Values that are simple single numbers are output as u32 constants.
  - Everything else is output as a string constant, with `_DEBUG` appended to the symbol name, and the string a debug dump of the value data structure of that property.

Also, this is built on top of #75904.

d3zd3z committed:
rust: Initial config option

Add the `CONFIG_RUST` Kconfig.  This can be set to indicate that an
application wishes to use Rust support.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Simple main wrapper

Until further variants are needed, this provides a `main()` function
that simply calls into the `rust_main()` function.  This is adequate for
applications where main can be directly written in Rust.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Basic `zephyr` crate

The initial support crate for zephyr use within a Rust application.
This crate does two simple things:

- Processes the .config file generated for this build, and ensures that
  CONFIG_RUST is enabled.
- Provide a hanging Rust panic handler.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Add cmake support for building rust apps

This provides the function `rust_cargo_application()` within Cmake to
give applications written in Rust a simple way to start.

This implements the basic functionality needed to build a rust
application on Arm, including target mapping, and invoking cargo during
the build.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Simple rust hello_world application

This implements a simple hello world application in Rust.  Since there
are no bindings yet, this just directly calls `printk` from Rust code.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
MAINTAINERS: Add Rust to maintainers file

Indicate the new code is maintained.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Move Kconfig bool values into crate

Create a crate `zephyr-build` that contains the code that will run at
build time.  Move the bool kconfig handing from this.  This will make it
easier for an application crate that needs access to config entries to
do access them by:

- Adding a dependency to zephyr-build to `[build-dependencies]` in their
  Cargo.toml.
- Creating a build.rs, or adding a call to:
  zephyr_build::export_bool_kconfig() in it.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Make Kconfig values available to Rust code

As part of the build of the `zephyr` crate, convert string and numeric
entries from the .config file of the current build into constants.  This
mimics how these are available as defines in C code.

The defines are available under `zephyr::kconfig::CONFIG_...`.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Access CONFIG_BOARD

This is a bit awkward, as we don't have a clean interface between Rust
and the Zephyr C code, but copy and null-terminate the current board,
and include that in the printed message.

This demonstrates that the kconfig module generation is working.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
doc: languages: Add Rust documentation

Add initial docs for Rust language support.  Explains the support
currently present, how to install the toolchain needed, and explains
what is in the current sample.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Rename hello

The documentation system requires all of the samples to have unique
names, no matter what directory they are in.  Rename our sample to
`hello_rust` to avoid a conflict with the existing `hello_world` sample.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
doc: Use proper quote block for shell

Code-blocks for shell commands are of type 'shell', to avoid this
warning.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Support negative Kconfig int values

Distinguish between 'int' and 'hex' Kconfig values.  Use `usize` for the
hex values, as these are unsigned and may have the high bit set.  But,
for 'int' values, use `isize`, and allow the constants to be negative.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Mark RUST support as experimental

This is still very experimental, mark it in the Kconfig as such.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
cmake: rust: Use proper FPU config

For Arm targets, match the ABI selection defines from the
cmake/compiler/gcc/target_arm.cmake file.  What is important here is
that the floating point selection is the same between gcc and rustc.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
doc: rust: Fix code block

Because of a mismatched quote, shell quoting for this block fails
sphinx.  Removing the quite causes compliance to fail due to the
mispelling.  Fix this by just removing formatting entirely from the
block.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
doc: rust: Use real config settings

Compliance fails if even an example refers to non-existent Kconfig
options.  Change these to real options, even though that is slightly
distracting.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
cmake: rust: Fix Cortex-M4 non-fp target

The Cortex-M4 should be the thumbv7em not thumbv7m.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Fix prototype for rust_main

Make this a proper C prototype function that takes no arguments.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Pass binary output dir to Rust

Rust build tools will need to be able to look at some generated files,
which are referenced from this directory.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: zephyr-build: Conversion of Device Tree

This code parses the DTS file generated by the Zephyr build, along with a
few entries from the generated header file, to build a representation of
the device tree.  It outputs a "correct" but only marginally useful
representation of this tree as a Rust module.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: zephyr: Use generated DT module

Invoke the DT generation, and include it at the top-level of the Zephyr
module.  With this change, clients can use module paths within Rust to
access Device tree fields.

   zephyr::devicetree::chosen::zephyr_console::OID

would be the OID field for the uart device for the console.  Currently,
most of the properties are encoded as debug strings, so this is of limited
usefulness yet.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Add allocator support

Create a config `CONFIG_RUST_ALLOC` that will hook Rust's allocation system
into the `malloc`/`free` allocator provided on Zephyr.  This will allow the
`alloc` crate in rust to be used.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Simplify to use allocation

Make use of allocation in the `rust_hello` sample to clean up the messy
invocation.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Provide a `println` macro when alloc is available

When the `CONFIG_RUST_ALLOC` has been set, provide a `println` macro that
can be called to print messages using `printk` in Zephyr.  This will need
to perform allocations equal to at least twice the size of the printed
message, that will be freed after the print.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Simplify the rust hello

With allocation enabled, use the new `println` macro provided in the
`zephyr` crate.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Messy hand-generated sys binding

Hand generate some bindings to enough sys entries to make GPIO calls.  This
is not intended to persist, but to determine what this will end up needing
to be.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Add blinky

Add a simple blinky port.  This hand-codes the DT entries for a single
nordic board, and these will end up coming from the DT.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Add augments to Rust DT generation

The Augment mechanism allows us to specify snippets of code that get added
to nodes in the device tree when converting to rust.  Adds support for
'gpio' nodes and the 'gpio-leds' children nodes, such that these nodes will
have a `get_instance()` function in them to return an instance of the gpio
pin driver that allows the pin to be manipulated.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: sys: Export the gpio based types

Make sure these are visible, so that the generated DT module can reference
them.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: sys: Make gpio methods require `mut`

Whether the Rust interfaces to devices on Zephyr should require `mut` is a
bit of a difficult issue.  Zephyr itself doesn't require them at
all (struct device is always passed as a const pointer).  In the C case,
that is likely because they live in ROM.  But, with the architecture we
have, Zephyr typically has multiple things that reference the same gpio
device, and requiring `mut` would make them almost unusable.

We'll make these `mut` for now, as they can be.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
sample: rust: blinky: Use DT instead of unsafe

Remove the unsafe code from the rust blinky sample, and use the instancing
directly from the generated devicetree.rs.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Remove constants from blinky

Now that these are defined in `zephyr::sys`, just use the defined constant
rather than hardcoded numbers.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
WIP

Snapshot of work in progress

d3zd3z committed:
cmake: rust: Fix typo

Small typo in a comment.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: zephyr-build: Fix logic

These should be elses.  The regexps shouldn't overlap, but this will
guarantee we don't produce redundant declarations.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Ignore Cargo.lock files

These are generated by cargo to capture the current dependencies.  Per the
cargo docs, they should either be checked in, or added to .gitignore.  We
will do the later so that dependencies are always resolved.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
lib: rust: Add wrapper for k_str_out

Until Rust has support for syscalls, add a C wrapper for `k_str_out`, so it
can be called from rust code to enable debug messages.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Implement `printk` and `printkln`

These macros provide debugging message support for Rust code, in the same
vein as printk.  These use the same console output that `printk` does from
C code.

This should be callable from userspace.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Use printkln macro

Remove the unsafe code calling into the C printk, instead using the new
`printkln` macro for emit a formatted message.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Work around rust's alloc and userspace

The Rust allocator invokes a global symbol
`__rust_no_alloc_shim_is_unstable`, that gets defined by the allocator
definition in an application.  When trying to use allocation in Rust
from userspace, the fetch from this location faults, as the `.bss`
section is not accessible to userspace programs.

Work around this by creating a symbol of the same name, and declaring it
a const, which places it in the readonly data, which is accessible.
This will override the segment destination of the generated code, and
allow the allocator to continue.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Only provide Rust on supported platforms

Implement a `RUST_SUPPORTED` Kconfig, and initially enable it for the
Cortex-M targets that are now supported.  This also allows samples and test
to be filtered based on the availability of Rust.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Remove RTT from the sample

In order for this to pass qemu tests, the console needs to come out of the
uart.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Enable Rust hello as sample

Use the `CONFIG_RUST_SUPPORTED` check to make sure this only tries
running on targets where Rust is supported.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
fix merge errors

d3zd3z committed:
rust: Add support for some RISC-V targets

Adds build support for Rust targets for the RV32I and RV64I ISA.  RV64I
targets require M, A and C extensions.  The 32-bit rust tools have a few
options, but for now, build with none of the extensions.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z commented:
Removed the alloc and println, as the main initial Rust PR adds a non-allocating printk/printkln.

d3zd3z committed:
rust: Ignore possible left files

It can be useful to put a .cargo/config.toml symlink in the lib/rust/zephyr
directory to allow the rust tools to get the current build configuration.
However, we don't want to commit this file, nor the Cargo.toml file that
gets left behind when doing this.

Also adds target/ in case the developer accidentally asks their IDE to
build the code.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Create zephyr-sys for low-level bindings

With a lot of work done by Mario Jaun <mario.jaun@zuehlke.com>.

Use the rust-bindgen tool to extract bindings from the Zephyr header files.
The tool is run with the exact configuration of the current build, and
therefore the bindings will match the current target.  The ifdefs and such
are not translated into the Rust code, and this must be generated live for
each build.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Use zephyr-sys to get bindings

Instead of trying to manually generate the bindings, use those generated by
bindgen in the zephyr-sys crate.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Convert k_str_out to bindgen one

Remove the manual k_str_out wrapper, and use the one generated by bindgen.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
include: Add dummy field for Rust as well as CPP

For CPP builds, a few structs that end up empty in some configurations will
generate compile errors.  With the Rust tools, bindgen ends up producing
empty structsw for these as well.  Although the code compiles, it generates
warnings.  For now, add CONFIG_RUST to the ifdef checks so that these
structs don't end up empty with Rust either.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
cmake: rust: Compile the bindgen wrapper

When compiling Rust programs, the bindgen utility generates a wrapper file
to expand inline functions, as these otherwise cannot be accessed from
rust.  Pass a consistent name to the cargo build, and add it to our source
build so that this wrapper get compiled.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Ignore unknown cfgs in the zephyr crate

We use the `#[cfg(...)]` directive in rust to pass Kconfig values through.
Because it is possible for the code to depend on a Kconfig value that isn't
present in a given build, there isn't a way for us to tell the Rust
toolchain about all possible config values.

For now, just disable this warning entirely.  This functionality could be
supported by the patch validation scripts, which seems like a better place
than trying to gather a list of all possible configs at build time.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
cmake: rust: Fix typo in cmake

Typo in symbol spelling.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Adapt DT to new zephyr-sys

Now that we have bindings available in zephyr-sys, adapt the GPIO DT
bindings ot use these.  This cleans up the hand-generated bindings we've
been using.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Add a simple rust logger using printk

Implement a log handler for the `log` crate that uses printk.  This
currently is always enabled (although if unused it doesn't add any code).

An application can invoke `zephyr::set_logger();` to install this log
handler, and messages logged with the standard logger will be printed using
the printk mechanism in Zephyr.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: blink: Adapt to newer code

Adjust the DT usage to work with the new code, and use the simple logging
infrastructure to log messages.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z commented:
I've rebased this on top of #76337, as having generated bindings makes a lot of this quite a bit easier.

d3zd3z committed:
rust: Work around clang not defining __SOFTFP__

Gcc automatically defines __SOFTFP__ on Cortex-M targets where soft
floating point is used.  Clang, which is used by bindgen, does not
define this.  Workaround this by detecting this situation, and defining
this.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Provide critical-section for Zephyr

Implement a general critical section handler, using spinlocks, for Rust
users to be able to use the `critical-section` crate.  This crate is
used by other crates as well.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Allow GpioPin to not be used

Whether a device uses a gpio pin depends on the DT.  Allow this symbol
to be unused, for targets that doing have gpio pins defined.

Signed-off-by: David Brown <david.brown@linaro.org>

pdgendt commented on samples/rust/blinky_rust/src/lib.rs:
```diff
@@ -0,0 +1,55 @@
+unsafe extern "C" fn blink(_p1: *mut c_void, _p2: *mut c_void, _p3: *mut c_void) {
+    warn!("Inside of blinky");
+
+    let mut led0 = zephyr::devicetree::aliases::led0::get_instance();
```
Can we somehow make use of `Option` types, similar to how `DEVICE_DT_GET_OR_NULL` would work? This would allow to re-use code for different platforms where runtime checks are forced if a device tree node would be available or not.

d3zd3z replied:
That seems reasonable, but I'll have to think about how that would work. I do like the idea of the DT being actually present as modules, but I'm not aware of a way to query that, and it doesn't really make sense to query the existence of modules at compile time.

d3zd3z committed:
rust: Add zephyr-sys into the Rust builds

This crate runs bindgen at build time to generate bindings to the Zephyr
API surface.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Add a kconfig RUST_CHECK_KOBJ_INIT

This adds an atomic to each static Rust-defined kobject to ensure that they
are properly initialized.  It adds a small amount of overhead to each
object, as well as to when they are first used.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: zephyr-sys: additional bindings

The bindgen tool will generate bindings for #defines that resolve to
numeric constants.  This misses some otherwise constant bindings that have
a type cast in them.

We can, however, export some of these, by making a const in the C code that
bindgen will export.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Add kobject support with threads

Implement a simple static wrapper to use with Zephyr kernel objects, and
use this to implement static threads.  This defines a static thread and a
static thread stack that can be initialized with a simple rust function,
with no arguments.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: philosopher

Start an implementation of the dining philosopher's problem, in Rust.
There is just a child thread, and the mutex use is unsafe.  But, this
demonstrates the new Rust thread support.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Support closure thread creation

When alloc is enabled for Rust code, provide an easier to use `spawn`
function that boxes a closure and sends it to the child side.  This
requires allocations to hold the data passed to the child.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Use `spawn` to start child threads

Change this to test the `spawn` function we have available with allocation.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Disable .noinit for now

Mark this as a TODO, disabling putting the k_thread in the noinit section.
It is supposed to only do this when we aren't debugging initialization, but
the conditional attribute doesn't seem to work.  For now, put it in
initialized memory, which will run correctly, just a little longer on init
time, but the atomic will be initialized properly.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Add a simple safe wrapper for k_mutex

`zephyr::sys::Mutex` is a simple wrapper for the k_mutex that provides safe
access to lock and unlock.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Change philosophers to use sys::Mutex

This eliminates the unsafe operations here.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
linker: Include the .ARM.extab section for Rust

This is a little complicated.  We currently just jump directly to rust's
panic, so we aren't likely to actually need these unwind tables.  But, they
would be needed if we ever reall paniced.  As of the time of this change,
the tables seem to just be 24 bytes of ROM space.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Implement sync::Mutex

This is a Mutex object, similar to that in Rust's std::sync::Mutex.  It
doesn't implement poisoning yet, as we have no poisoning to speak of.

This can be used as an `Arc<Mutex<T>>` to protect a block of data.

The initializer requires a sys::Mutex to be passed in, as these are
typically static in Rust applications.  The wrapped mutex can live in an
Arc.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Put K_FOREVER and K_NO_WAIT in sys

Make these symbols available to Rust code.  They are too complex for
bindgen to inline, so represent the values as consts, which does depend on
their specific values.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
rust: Allow sys::Mutex to be cloned

The low-level mutex can be safely cloned (and is send as well).  Be sure to
derive clone so that it can be used this way.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z committed:
samples: rust: Use sync::Mutex for philosopher demo

The philosopher demo now uses the "friendly" wrapped mutexes, and as part
of the demo, places an incrementing counter inside of the mutex.

Signed-off-by: David Brown <david.brown@linaro.org>

d3zd3z commented:
Rust on Zephyr work is happening on the zephyr-lang-rust module. This is no longer relevant.
